vowpalwabbit/cache.cc

```diff
@@ -70,98 +70,102 @@ void VW::write_example_to_cache(io_buf& output, example* ae, label_parser& lbl_p
 {
-  // Unused for now.
-  uint64_t size;
-  char* read_ptr;
-  if (input.buf_read(read_ptr, sizeof(size)) < sizeof(size)) { return 0; }
-  memcpy(&size, read_ptr, sizeof(size));
-
-  ae->sorted = sorted_cache;
-  size_t total = lbl_parser.read_cached_label(ae->l, ae->_reduction_features, input);
-  if (total == 0) { return 0; }
-  if (read_cached_tag(input, ae) == 0) { return 0; }
-  unsigned char newline_indicator = input.read_value<unsigned char>("newline_indicator");
-  if (newline_indicator == newline_example) { ae->is_newline = true; }
-  else
+  size_t temp;
+  if ((temp = input.buf_read(c, sizeof(index) + sizeof(size_t))) < sizeof(index) + sizeof(size_t))
   {
-    ae->is_newline = false;
+    VW::io::logger::errlog_error("truncated example! {} {} ", temp, char_size + sizeof(size_t));
+    return 0;
   }
+  index = *reinterpret_cast<unsigned char*>(c);
+  c += sizeof(index);
+  return sizeof(index);
+}
 
-  // read indices
-  unsigned char num_indices = input.read_value<unsigned char>("num_indices");
-
-  char* c;
-  for (; num_indices > 0; num_indices--)
+size_t read_cached_features(io_buf& input, features& ours, bool& sorted, char*& c)
+{
+  size_t total = 0;
+  size_t storage = *reinterpret_cast<size_t*>(c);
+  c += sizeof(size_t);
+  input.set(c);
+  total += storage;
+  if (input.buf_read(c, storage) < storage)
   {
-    size_t temp;
-    unsigned char index = 0;
-    if ((temp = input.buf_read(c, sizeof(index) + sizeof(size_t))) < sizeof(index) + sizeof(size_t))
-    {
-      VW::io::logger::errlog_error("truncated example! {} {} ", temp, char_size + sizeof(size_t));
-      return 0;
-    }
+    VW::io::logger::errlog_error("truncated example! wanted: {} bytes ", storage);
+    return total;
+  }
 
-    index = *reinterpret_cast<unsigned char*>(c);
-    c += sizeof(index);
+  char* end = c + storage;
+  uint64_t last = 0;
 
-    ae->indices.push_back(static_cast<size_t>(index));
-    features& ours = ae->feature_space[index];
-    size_t storage = *reinterpret_cast<size_t*>(c);
-    c += sizeof(size_t);
-    input.set(c);
-    total += storage;
-    if (input.buf_read(c, storage) < storage)
+  for (; c != end;)
+  {
+    feature_index i = 0;
+    c = run_len_decode(c, i);
+    feature_value v = 1.f;
+    if (i & neg_1)
+      v = -1.;
+    else if (i & general)
     {
-      VW::io::logger::errlog_error("truncated example! wanted: {} bytes ", storage);
-      return 0;
+      v = (reinterpret_cast<one_float*>(c))->f;
+      c += sizeof(float);
     }
+    uint64_t diff = i >> 2;
+    int64_t s_diff = ZigZagDecode(diff);
+    if (s_diff < 0) { sorted = false; }
+    i = last + s_diff;
+    last = i;
+    ours.push_back(v, i);
+  }
+  input.set(c);
+  return total;
+}
 
-    char* end = c + storage;
+int VW::read_example_from_cache(VW::workspace* all, io_buf& input, v_array<example*>& examples)
+{
+  char* read_ptr;
+  if (input.buf_read(read_ptr, sizeof(uint64_t)) < sizeof(uint64_t)) { return 0; }
```

Comment:
Even though `size` had an "unused for now" comment, can you please leave a comment to describe what the value is. This was added to be able to skip cached examples on a read failure and is intended to be used.

Reply:
Readded a comment with `size`